---
title: "Annotation plots with smallCAGEqc"
author: "Charles Plessy"
date: "`r Sys.Date()`"
output:
  html_document: 
    keep_md: yes
    toc: yes
  rmarkdown::html_vignette:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Annotation plots with smallCAGEqc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

R commands
==========

The main command to produce annotation plots in _smallCAGEqc_ is called
`plotAnnot`.  It takes a table containing the sample metadata, a _scope_,
a title and optionally a factor to group similar plots together.

The _scope_ determines what data is plotted and how it is normalised.
The available scopes will be explained with an example plot in a later part
of this document, but first, let's see the input in more details.

Here, we will use some of the example data that is distributed in _smallCAGEqc_.
The commands below load the R package and load the example data in a data frame
called `libs`.

```{r libs, message = FALSE}
library(smallCAGEqc)
libs <- read.table(system.file("extdata/libs-with-all-metadata.tsv", package="smallCAGEqc"))
```


Sample metadata
---------------

The following columns in the metadata table describe the total remaining pairs
step after step in the processing.

 - **total**: The total number of pairs before tag extraction (called "raw" in
   some other pipelines). In cases where this number is not available per sample,
   for example when demultiplexing and tag extraction are performed at the same
   stage, it is set arbitrarily to zero.

 - **extracted**: The number of pairs where the linkers and unique molecular
   identifier (if present) were successfully extracted. 
   
 - **cleaned**: The number of pairs remaining after filtering out spike, rRNA,
   low-complexity, primer artefact and other unwanted sequences.
   
 - **mapped**: The number of pairs with at least one successful alignment. This
   called "genome_mapped" in some other pipelines.
 
 - **properpairs**: The number or pair remaining after filtering out the
   non-proper alignments.  This is called "properly_mapped" in some other
   pipelines.
 
 - **counts**: The number of unique molecules counted after alignment.  This is
   called "transcript_count" in some other pipelines.
 
The following columns describe the number of pairs removed at some step of
the processing.

 - **spikes**, **rdna**: The number of pairs removed because they matched
   spikes or rRNA reference sequences, respectively.
  
 - **tagdust**: The number of pairs removed because of low-complexity or
   similarity to primer artefacts.
   
The following columns describe the number of TSS (after proper pairing and
deduplication) aligning to known regions in the genome.

 - **promoter**: Promoter regions.
 
 - **exon**: Known exons.
 
 - **intron**: Know introns (that is: the TSS matches a transcript but none of
   its exons).
   
 - **unknown**: None of the above; basically intergenic.
 
 - **other**: Normally there should not be anything in this category since
   _unknown_ should catch all intergenic TSS.  But sometimes there is a
   misdesign, for instance in the example data here, the annotation file
   contained one more category, "boundary", which was not handled correctly
   by smallCAGEqc.
   
The annotation is hierarchical (promoters have priority on exons, etc.), so
the sum of the annotation columns above should be be equal to the _counts_
column.


Different types of scopes
=========================

Step-by-step extraction
-----------------------

Shows how many pairs are removed by the extraction, cleaning, mapping
(proper pairs) and transcript counting steps described above.

 - **Extraction**: The total number of pairs where a tag could not be extracted
   (_extracted_ - _cleaned_).
   
 - **Cleaning**: The total number of pairs where a tag was discarded because
   it matched a reference or artefact sequence, or because it had low complexity
   (_tagdust_ + _rdna_ + _spikes_, or _cleaned_ - _mapped_).
   
 - **Mapping**: The number of non-mapped pairs (_mapped_ - _properpairs_).
 
 - **Deduplication**: The number of non-properly mapped pairs (_properpairs_ -
   _counts_).
 
 - **Counts**: The number of molecule _counts_.
 

```{r steps}
plotAnnot(libs, "steps", "steps")
```

QC report
---------

Pairs are categorised as tag dust, rDNA, spikes, unmapped, non-proper,
duplicates and counts, and normalised by the total number of extracted pairs.
Non-extracted pairs are ignored.

Compared to "steps", this scope gives more details on the sequences removed
at the TagDust and mapping stages of the processing pipeline.

 - **Tag_dust**, **rDNA**, **Spikes**: same numbers as in the _tagdust_, _rdna_
   and _spikes_ columns of the metadata table.
   
 - **Unmapped**: The number of pairs that could not be mapped.
 
 - **Non_proper**: The number of pairs that did not have a _proper_ alignment.

 - **Duplicates**: The number of pairs that do not add a molecule count.
 
 - **Counts**: same as in "steps".


```{r qc}
plotAnnot(libs, "qc", "qc")
```


Annotation of the transcript or tag counts
------------------------------------------

The unique molecule counts are grouped in annotation categories
("promoter", "exon", "intron" and "intergenic"), as described above.

```{r counts}
plotAnnot(libs[libs$counts > 0,], "counts", "counts")
```


Annotation of the mapped reads
------------------------------

Same as "counts", with the addition of duplicates and non-proper pairs.
Therefore the plot represents all the mapped data.

```{r mapped}
plotAnnot(libs, "mapped", "mapped")
```


QC including annotations
------------------------

Pairs are categorised by extraction step and genome annotation.

```{r all}
plotAnnot(libs, "all", "all")
```


Annotation, normalised by mapped reads
--------------------------------------

Same as `all` except that normalisation is relative to the number of mapped reads

```{r annotation}
plotAnnot(libs, "annotation", "annotation")
```


Contents of the `libs` table
============================

```{r show-libs}
libs
```

R session info
==============

```{r session-info}
sessionInfo()
```
