---
title: "<Insert title>"
output: 
  html_document: 
    keep_md: yes
    number_sections: yes
    toc: yes
---

```{r echo=FALSE}
knitr::opts_chunk$set(cache  = TRUE)
knitr::opts_knit$set(verbose = TRUE)
# knitr::opts_chunk$set(dev=c('svg', 'png'))
options(width = 120)
ggplot2::theme_set(ggplot2::theme_bw())
```

```{r library_information_1}
LIBRARY      <- "<Insert library ID>"  # For example, the sequencing run ID.
ZENBU_COLLAB <- "<Insert Zenbu collaboration ID>"
WORKFLOW     <- "<Insert workflow ID>" # For example "OP-WORKFLOW-CAGEscan-short-reads-v2.0"
MOIRAI_STAMP <- "<Insert Moirai run stamp>" # For example "20150803121351"
MOIRAI_PROJ  <- "<Insert Moirai user name>" # For example "Charles"
MOIRAI_USER  <- "<Insert Moirai user name>" # For example "nano-fluidigm"
ASSEMBLY     <- "<Insert name of genome assembly>" # For example "hg38"

# The following is still hardcoded...
# See https://gist.github.com/charles-plessy/9dbc8bc98fb773bf71b6 for hints

GROUP_SHARED <- "/osc-fs_home/scratch/gmtu"
GENE_SYMBOLS <- paste0(GROUP_SHARED, "/annotation/homo_sapiens/gencode-23/gencode.v23.annotation.genes.bed")
ANNOTATION   <- paste0(GROUP_SHARED, "/annotation/homo_sapiens/gencode-23/gencode.v23.annotation.bed")

# In case of a pair of Fludigm C1 CAGE runs, uncomment the lines below and fill
# insert missing information.  Otherwise, delete or leave as it is.
#RunA         <- "<Insert ID of Run A>"
#RunB         <- "<Insert ID of Run B>"
#ctrls        <- list( RunA=list(posi="<Insert well ID>", nega="<Insert well ID>")
#                    , RunB=list(posi="<Insert well ID>", nega="<Insert well ID>"))
```

<!-- Example title: MiSeq run `r LIBRARY` -->

<Insert title>
==============

 - C1 runs `r RunA` and `r RunB`.
 - Sequenced on MiSeq (`r LIBRARY`)
 
<!-- The lines about cellomics and controls below can be deleted if not using
     C1 CAGE -->

- cellomics IDs:
   - `r RunA`: <Insert short description of Run A>
   - `r RunB`: <Insert short description of Run B>

 - Positive control (200 cells): `r ctrls$RunA$posi`, `r ctrls$RunB$posi`.
 - Negative control: `r ctrls$RunA$nega`, `r ctrls$RunB$nega`.
 
Jump directly to the [Analysis](#analysis) section if you are not interested in the details of the data processing.


Data Processing
===============

Link files from MOIRAI.
-----------------------

```{r library_information_2}
BASEDIR      <- "/osc-fs_home/scratch/moirai"
MOIRAI_RUN   <- paste(LIBRARY, WORKFLOW, MOIRAI_STAMP, sep=".")
MOIRAI_BASE  <- paste(BASEDIR, MOIRAI_USER, "project", MOIRAI_PROJ, MOIRAI_RUN, sep="/")
MOIRAI_ID    <- paste(LIBRARY, WORKFLOW, MOIRAI_STAMP, sep=".")
MOIRAI_FRAGS <- paste0(MOIRAI_BASE, "/CAGEscan_fragments")
MOIRAI_BAM   <- paste0(MOIRAI_BASE, "/genome_mapped")
MOIRAI_URL   <- paste0("http://moirai.gsc.riken.jp/", MOIRAI_BASE, "/", MOIRAI_RUN, ".html")

smallCAGEqc::exportInEnv("LIBRARY", "MOIRAI_BASE", "GENE_SYMBOLS", "ANNOTATION")
```

 Intranet link: [`r MOIRAI_RUN`](`r MOIRAI_URL`)

Create a symbolic link from Moirai results to the working directory.

```{r create_link, engine="bash"}
[ -L $LIBRARY.l1.osc.gz ] ||
  ln --symbolic $MOIRAI_BASE/$LIBRARY.l1.osc.gz
```

Annotation and gene symbols.
----------------------------

```{r annotate_bed, engine="bash"}
function osc2bed {
  zcat $1 |
    grep -v \# |
    sed 1d |
    awk '{OFS="\t"}{print $2, $3, $4, "l1", "1000", $5}'
}

function bed2annot {
  bedtools intersect -a $1 -b $ANNOTATION -s -loj |
    awk '{OFS="\t"}{print $1":"$2"-"$3$6,$10}' | 
    bedtools groupby -g 1 -c 2 -o collapse
}

function bed2symbols {
  bedtools intersect -a $1 -b $GENE_SYMBOLS -s -loj |
    awk '{OFS="\t"}{print $1":"$2"-"$3$6,$10}' | 
    bedtools groupby -g 1 -c 2 -o distinct
}

osc2bed $LIBRARY.l1.osc.gz | tee $LIBRARY.l1.bed | bed2annot - > $LIBRARY.l1.annot
bed2symbols $LIBRARY.l1.bed > $LIBRARY.l1.genes
```

Data load and QC in R.
======================

Moirai metadata
---------------

```{r load_libs_and_moirai, message=F}
library(oscR)
library(smallCAGEqc)
stopifnot(
    packageVersion("oscR") >= "0.2.0"
  , packageVersion("smallCAGEqc") > "0.11.2"
)
library(data.table)
library(magrittr)
library(reshape)
library(ggplot2)
library(vegan)
library(plyr)

PROCESSED_DATA <- MOIRAI_BASE

libs <- loadMoiraiStats() #  %>% llPostProcess('nano-fluidigm')
```

Transcript counts (properly paired)
-----------------------------------

Note that `fread` needs to temporarly save an uncompressed copy of the file
that is being read.  If there is not enough space in the temporary directory
(often `\tmp`),  either uncompress the file first, or point to a larger
directory to contain the temporary files (for example, run `R` with `TMPDIR=.
R`).

Note that _data.tables_ seem to be saved as simple _data frames_ when
closing a session.  After restoring a workspace, one needs to re-convert with the `data.table` command.

```{r load_osc_data}
l1 <- fread.osc(paste0(LIBRARY, ".l1.osc.gz"), dropIdCoords=TRUE)

setnames(l1, colnames(l1) %>% sub('raw.', '', .) %>% sub('.None', '', .) %>% sub('RunA',RunA,.) %>% sub('RunB',RunB,.))

# Be sure that the "libs" and "l1" tables have matching names.
stopifnot(all(colnames(l1) == rownames(libs)))
```

Annotation
----------

```{r annotate_l1}
annot.l1 <- read.table(paste(LIBRARY,'l1','annot',sep='.'), head=F, col.names=c('id', 'feature'), row.names=1)
annot.l1 <- hierarchAnnot(annot.l1)

libs <- cbind(libs, t(rowsum(l1,  annot.l1[,'class']))[rownames(libs),])
```

Gene count
----------

```{r genesymbols_l1}
genesymbols <- read.table(paste(LIBRARY,'l1','genes',sep='.'), col.names=c("cluster","symbol"), stringsAsFactors=FALSE)
rownames(genesymbols) <- genesymbols$cluster

genes <- rowsum(l1, genesymbols$symbol)

libs$genes       <- colSums(genes > 0)
libs$geneSymbols <- countSymbols(genes) # Not the same!
```


Raw reads per molecule (BED12 data)
-----------------------------------

```{r load_fragments}
bed <- loadBED12( libs$samplename %>%
             sub(RunA, "RunA", .) %>%
             sub(RunB, "RunB", .) %>%
             paste0(MOIRAI_FRAGS, "/", ., ".bed"))
```

Cell pictures
-------------

```{r load_fluo_data}
read.fluo <- function(RUN) read.delim( paste0("../imageJ/", RUN, ".txt")
                                     , row.names="cell_id"
                                     , stringsAsFactors = FALSE)
fluo <- rbind(read.fluo(RunA), read.fluo(RunB))
libs <- cbind(libs, subset(fluo,,c("mean_ch2", "bg_mean_ch2", "mean_ch3", "bg_mean_ch3", "Error", "Comment")))
libs$Error <- factor( libs$Error
                    , levels=c("0-No Error", "1-No cell", "2-Debris", "3-OutOfFocus", "4-MultipleCells", "5-Control", "6-Dead"))
```

A hardocoded treshold of 2.5 is used to identify dead cells.  The histogram
below is to check if this value makes sense in this dataset.

```{r dead-cells, dev=c('svg', 'png'), fig.height=2.5}
hist( libs$mean_ch3 - libs$bg_mean_ch3
    , br   = 100
    , main = "Current threshold for identifying dead cells"
    , xlab = "mean_ch3 - bg_mean_ch3" )
deadThresh <- 2.5
abline(v=deadThresh, col="red")

libs[libs$mean_ch3 - libs$bg_mean_ch3 > deadThresh, "Error"] <- "6-Dead"
```

Controls
--------

Some samples with errors were repalced by the positive and negative controls.

```{r flag-controls}
libs[libs$Well %in% ctrls[["RunA"]] & libs$Run == RunA, "Error"]   <- "5-Control"
libs[libs$Well %in% ctrls[["RunB"]] & libs$Run == RunB, "Error"]   <- "5-Control"
libs[libs$Well == ctrls$RunA$posi & libs$Run == RunA, "Comment"] <- "Positive control"
libs[libs$Well == ctrls$RunB$posi & libs$Run == RunB, "Comment"] <- "Positive control"
libs[libs$Well == ctrls$RunA$nega & libs$Run == RunA, "Comment"] <- "Negative control"
libs[libs$Well == ctrls$RunB$nega & libs$Run == RunB, "Comment"] <- "Negative control"
```

cDNA concentration.
-------------------

```{r cDNA_concentration, dev=c('svg', 'png'), fig.height=2.5, message=FALSE, warning=FALSE}
read.pg <- function(RUN)
  paste0("../cDNA_yields/", RUN, ".picogreen.xlsx") %>%
    fldgmPicoGreen("PN 100-6160") %>%
    extract(,"concentration")
libs$Concentration <- c(read.pg(RunA), read.pg(RunB))
fldgmConcentrationPlot(libs)
```

Combined analysis of fluorescence and cDNA concentration.
---------------------------------------------------------

### Array heatmaps.

```{r define_fldgmArrayQCplot}
fldgmArrayQCplot <- function(RUN) fldgmArrayQC(libs[libs$Run==RUN,], RUN)
```

```{r 'runA.arrayQC', dev=c('svg', 'png'), fig.height=2.5}
fldgmArrayQCplot(RunA)
```

```{r 'runB.arrayQC', dev=c('svg', 'png'), fig.height=2.5}
fldgmArrayQCplot(RunB)
```

### Live / dead stain and DNA concentration.

Both runs are plotted together; this explains why there may be two groups, when cDNA
yields differed strongly. 

```{r live-dead, dev=c('svg', 'png')}
with(libs, plotNameColor(Concentration, mean_ch2 - bg_mean_ch2, Error, Well))
with(libs, plotNameColor(Concentration, mean_ch3 - bg_mean_ch3, Error, Well))
```

Richness
--------

```{r calculate-richness}
libs$l1 <- colSums(l1 > 0)
libs$r100l1 <- rarefy(t(l1),100)
```

Analysis
========

```{r richness-concentration, dev=c('svg', 'png')}
with(libs, plotNameColor(Concentration, r100l1, Error, Well))
```

```{r richness-outliers, dev=c('svg', 'png')}
with(subset(libs, Error == "0-No Error"), plotNameColor(Concentration, r100l1, Run, Well))
```

QC barplots
-----------

```{r qc-barplots, dev=c('svg', 'png')}
plotAnnot(libs, 'qc', LIBRARY, libs$Group)
```

Annotation
----------

```{r annotation-barplots, dev=c('svg', 'png')}
plotAnnot(libs, 'counts', LIBRARY, libs$Group)
```

Correlation between runs
------------------------

```{r correlation-heatmap, dev=c('svg', 'png')}
library(dendextend)

# Filtering samples with extraction count < nbFilterExtracted
# Note that it can also remove negative controls
nbFilterExtracted <- 10000

# Possibility to color the clusters dendrogram if nbClusters != 0
nbClusters <- 1

# Please check that length(colorsDend) >= nbClusters
colorsDend = c("black", "red", "darkgreen", "darkseagreen", "blue", "green", "darkorange3", "grey40", "darkorchid4", "pink3", "turquoise4")

filterExtracted <- libs$counts > nbFilterExtracted

corMat <- genes[-1, filterExtracted] %>% TPM %>% log1p %>% cor

corDend <-
  corMat %>%
  subtract(1, .) %>%  # Otherwise it is not a mathematical distance !
  as.dist %>% hclust %>% as.dendrogram %>%
  set("branches_k_color", value = colorsDend[1:nbClusters], k = nbClusters) %>%
  set("labels_colors", value = colorsDend[1:nbClusters], k = nbClusters) %>%
  set("labels_cex", .6) %>%
  ladderize(FALSE) # To ensure a reproducible visual order of the clusters

heatmapGenes <- NMF::aheatmap( corMat,
  annCol=list(Run=libs$Run[filterExtracted], Group=libs$group[filterExtracted]),
  Rowv = corDend, Colv = corDend )

# Dendrogram only: for visual comfort
(corDendGg <- corDend %>% as.ggdend %>% ggplot(horiz = F))
```

```{r correlation-heatmap-noerrors, dev=c('svg', 'png')}
# Uncomment if you have a Error column
#NMF::aheatmap( cor(genes[-1, libs$Error == "0-No Error"])
#             , annCol=list(Run=libs[libs$Error == "0-No Error", "Run"]))
```

```{r genes-dend-nb-clusters, dev=c('png','svg')}
if (nbClusters != 0) {
  # Need to create an hclust object to associate a cluster number to each cell
  genesDend    <- hclust(dist(cor(log(genes[-1, filterExtracted]+1))))
  genesDendRev <- rev(genesDend$height)
  
  # This plot can help user checking the number of clusters to choose
  plot(genesDendRev,
       col = c(colorsDend[1:nbClusters], rep("black", length(genesDendRev)-1)),
       pch = 20, ylab = "Clusters height")
  abline(v = nbClusters + 0.5, col = "grey60", lty = 2)
  
  # Dividing into nbClusters clusters
  genesDendClusters <- cutree(genesDend, nbClusters)
}
```

PCA analysis
------------

### PCA on Genes

```{r genes-pca}
library(ade4)
library(adegraphics)
library(dplyr)

# Normalisation in tags per million and removing the • genes
genesTpm <- t(TPM(genes[, filterExtracted]))[, -1]

# User can choose the number of axis to keep in the PCA
nbPcaAxis <- 3

genesPca <- dudi.pca(genesTpm, center = T, scale = F, scannf = F, nf = nbPcaAxis)
```

```{r genes-pca-summary, dev=c('png','svg')}
barplot(genesPca$eig, xlab = "Eigenvalues")
summary(genesPca)
```

```{r genes-pca-list-barplot, dev=c('png','svg')}
# List of genes ordered by components keeping information on 3 components
genesCoComp1 <- pcaCompGenesList(genesPca$co, 1)
genesCoComp2 <- pcaCompGenesList(genesPca$co, 2)
genesCoComp3 <- pcaCompGenesList(genesPca$co, 3)

# User can choose how many genes to display ?
nbDispGenes <- 25

genesCoComp1 %>% head(nbDispGenes) 
genesCoComp1 %>% tail(nbDispGenes)
genesCoComp2 %>% head(nbDispGenes) 
genesCoComp2 %>% tail(nbDispGenes)
genesCoComp3 %>% head(nbDispGenes) 
genesCoComp3 %>% tail(nbDispGenes)

plotHTB(genesCoComp1, 1, nbDispGenes)
plotHTB(genesCoComp2, 2, nbDispGenes)
plotHTB(genesCoComp3, 3, nbDispGenes)
```

```{r genes-pca-lico, dev=c('png','svg')}
# =================================================================
#         Clusters colored according to the dendrogram
# =================================================================
# /!\ step to be automatised /!\ 
#
# But for now needs manual calibration:
#
#   1 - Display 'genesDendClusters'
#   2 - Please match the cluster number and
#     the position of the color in 'colorsPca$color' (below)
#     You can type 'corDendGg' to display the dendrogram
#   3 - Check that you have 'length(colorsPca$color) == nbClusters'
#
# =================================================================

# User can change plot specificities
size         <- 1
transparency <- 0.6

if (nbClusters != 0) {
  colorsPca <- data_frame(group = 1:nbClusters,
    color = c("red", "darkgreen", "darkseagreen", "blue", "green", "darkorange3", "grey40", "darkorchid4", "pink3", "turquoise4", "black"))
  
  colorsPcaLi <- sapply(genesDendClusters, function(X) colorsPca$color[X])
  genesPlot <- pcaPlots(genesPca, colorsPcaLi, size, transparency)
} else {
  genesPlot <- pcaPlots(genesPca, size = size, transparency = transparency)
}

ADEgS(list(genesPlot$genes$ax12, genesPlot$genes$ax32, genesPlot$genes$ax13,
           genesPlot$cells$ax12, genesPlot$cells$ax32, genesPlot$cells$ax13),
  layout = c(2,3))
```

### Hints for the component biological interpretation

#### Nuclear enrichment ?

```{r nuclear-bash, engine="bash", eval=TRUE}
# NOTICE that geneSymbolFDRBestScoreNuclear.dat could be made once for all ...

bedtools intersect -a 2016.02.24.nuc.cyto.simon.bed.gz -b /osc-fs_home/scratch/gmtu/annotation/homo_sapiens/gencode-18/gencode.v18.annotation.genes.bed -s -loj | awk '{OFS="\t"}{print $1,$2,$3,$28,$5,$6,$7,$8}' | sort -k 4 | tee nuc.cyto.simon.annotated.bed | bedtools groupby -g 4 -c 5 -o max > listOfmaxValues.temp

join -a1 -e1 -1 1 -2 4  listOfmaxValues.temp nuc.cyto.simon.annotated.bed | awk '{OFS="\t"}{ if ($2==$6) print($1, $9) }' > geneSymbolFDRBestScoreNuclear.dat
```

```{r nuclear, dev=c('svg', 'png')}
library(gridExtra)
library(tidyr)

# Need to load dataset with nuclear score
nuclearFdr           <- read.table(file = "geneSymbolFDRBestScoreNuclear.dat", header = F)
colnames(nuclearFdr) <- c("geneNames", "FDR")

# User should choose a proper threshold
filterNuclearFdr <- 0.05

# Checking for duplicated genes, correction if needed
if (length(nuclearFdr$geneNames[duplicated(nuclearFdr$geneNames)]) != 0) 
  nuclearFdr %<>% group_by(geneNames) %>% 
                  summarise(FDR = min(FDR))

paste("##", nrow(nuclearFdr), "genes are referenced in geneSymbolFDRBestScoreNuclear.dat", sep = " ")
paste("##", sum(nuclearFdr$FDR < filterNuclearFdr), "genes are under the nuclear FDR threshold (", filterNuclearFdr, ")", sep = " ")

# Uniformisation of gene names for a better matching when joining
genesCoComp1$geneNames %<>% toupper()
genesCoComp2$geneNames %<>% toupper()
genesCoComp3$geneNames %<>% toupper()
nuclearFdr$geneNames   %<>% toupper()

genesCoComp1Fdr <- left_join(genesCoComp1, nuclearFdr)[,-6] %>% replace_na(list(FDR = 1))
genesCoComp2Fdr <- left_join(genesCoComp2, nuclearFdr)[,-6] %>% replace_na(list(FDR = 1))
genesCoComp3Fdr <- left_join(genesCoComp3, nuclearFdr)[,-6] %>% replace_na(list(FDR = 1))

# To recognize not referenced genes in nuclearPlot (*)
genesCoComp1Fdr$Referenced <- sapply(genesCoComp1Fdr$FDR, 
  function(X) ifelse(X==1, return(FALSE), return(TRUE)))
genesCoComp2Fdr$Referenced <- sapply(genesCoComp2Fdr$FDR, 
  function(X) ifelse(X==1, return(FALSE), return(TRUE)))
genesCoComp3Fdr$Referenced <- sapply(genesCoComp3Fdr$FDR, 
  function(X) ifelse(X==1, return(FALSE), return(TRUE)))

paste("##", round((nrow(genesCoComp1) - sum(genesCoComp1Fdr$Referenced==TRUE)) * 100 / nrow(genesCoComp1), 2), "% of the genes were not found in geneSymbolFDRBestScoreNuclear.dat", sep = " ")

nuclearPlotComp1 <- ggplot(genesCoComp1Fdr) + 
  geom_point(aes(x=Comp1, y=log(FDR), colour = FDR, shape = Referenced)) +
  scale_shape_manual(values=c(8, 20)) + 
  scale_color_continuous(low = "red", high = "black") + theme_classic() +
  geom_hline(aes(yintercept=log(filterNuclearFdr)), col = "red")
nuclearPlotComp2 <- ggplot(genesCoComp2Fdr) + 
  geom_point(aes(x=Comp2, y=log(FDR), colour = FDR, shape = Referenced)) +
  scale_shape_manual(values=c(8, 20)) + 
  scale_color_continuous(low = "red", high = "black") + theme_classic() +
  geom_hline(aes(yintercept=log(filterNuclearFdr)), col = "red")
nuclearPlotComp3 <- ggplot(genesCoComp3Fdr) +
  geom_point(aes(x=Comp3, y=log(FDR), colour = FDR, shape = Referenced)) +
  scale_shape_manual(values=c(8, 20)) + 
  scale_color_continuous(low = "red", high = "black") + theme_classic() +
  geom_hline(aes(yintercept=log(filterNuclearFdr)), col = "red")

grid.arrange(nuclearPlotComp1, nuclearPlotComp2, nuclearPlotComp3)
```

#### Bound to rough endoplasmic reticulum-associated ribosomes ?

```{r rER, dev=c('svg', 'png')}
# Loading and filtering supplemental material table S3
# from http://genome.cshlp.org/content/24/8/1396/suppl/DC1
ratRerSuppS3  <- read.csv(file = "Supplemental_Table.S3.csv", header = T)
ratRer        <- ratRerSuppS3 %>% select(Symbol.mouse, logFC.mbcyt, FDR.mbcyt)
names(ratRer) <- c("geneNames", "logFC", "FDR")

# Copying because of an issue. Explained in 2 comments...
ratRerOriginal <- ratRer

# Checking for duplicated genes, correction if needed
if (length(ratRer$geneNames[duplicated(ratRer$geneNames)]) != 0) 
  ratRer %<>% group_by(geneNames) %>% 
              summarise(FDR = min(FDR))

# Loosing logFC column so we have to merge it with the previous table
ratRer <- merge(ratRer, ratRerOriginal, by = c("geneNames", "FDR"))

# User should choose proper thresholds
filterLogFc  <- 3
filterRerFdr <- 0.05

ratRer$Filter <- sapply(ratRer$logFC, function(X) ifelse(X > filterLogFc, return(paste("logFC >", filterLogFc)), return(paste("logFC <", filterLogFc))))
# there is still duplicates... for now I ignore them. It should take the first value

paste("##", nrow(ratRer), "genes are referenced in Supplemental_Table.S3.csv", sep = " ")
paste("##", sum(ratRer$FDR < filterRerFdr), "genes are under the rER FDR threshold (", filterRerFdr, ")", sep = " ")
paste("##", sum(ratRer$logFC > filterLogFc), "genes are above the logFC threshold (", filterLogFc, ")", sep = " ")
paste("##", sum(ratRer$FDR[ratRer$logFC > filterLogFc] < filterRerFdr), "genes are under the rER FDR threshold (", filterRerFdr, ") and are above the logFC threshold (", filterLogFc, ")", sep = " ")

# Uniformisation of gene names for a better matching when joining
genesCoComp1$geneNames %<>% toupper()
genesCoComp2$geneNames %<>% toupper()
genesCoComp3$geneNames %<>% toupper()
ratRer$geneNames       %<>% toupper()

genesCoComp1RerFdr <- left_join(genesCoComp1, ratRer) %>% replace_na(list(FDR = 1))
genesCoComp2RerFdr <- left_join(genesCoComp2, ratRer) %>% replace_na(list(FDR = 1))
genesCoComp3RerFdr <- left_join(genesCoComp3, ratRer) %>% replace_na(list(FDR = 1))

# To recognize not referenced genes in rerPlot (*)
genesCoComp1RerFdr$Referenced <- sapply(genesCoComp1RerFdr$FDR, function(X) ifelse(X==1, return(FALSE), return(TRUE)))
genesCoComp2RerFdr$Referenced <- sapply(genesCoComp2RerFdr$FDR, function(X) ifelse(X==1, return(FALSE), return(TRUE)))
genesCoComp3RerFdr$Referenced <- sapply(genesCoComp3RerFdr$FDR, function(X) ifelse(X==1, return(FALSE), return(TRUE)))

paste("##", round((nrow(genesCoComp1) - sum(genesCoComp1RerFdr$Referenced==TRUE)) * 100 / nrow(genesCoComp1), 2), "% of the genes were not found in Supplemental_Table.S3.csv", sep = " ")

rerPlotComp1 <- ggplot(genesCoComp1RerFdr) +
  geom_point(aes(x=Comp1, y=log(FDR), colour = Filter, shape = Referenced)) +
  scale_shape_manual(values=c(8, 20)) +
  theme_classic() +
  geom_hline(aes(yintercept=log(filterRerFdr)), col = "red")
rerPlotComp2 <- ggplot(genesCoComp2RerFdr) +
  geom_point(aes(x=Comp2, y=log(FDR), colour = Filter, shape = Referenced)) +
  scale_shape_manual(values=c(8, 20)) + theme_classic() +
  geom_hline(aes(yintercept=log(filterRerFdr)), col = "red")
rerPlotComp3 <- ggplot(genesCoComp3RerFdr) +
  geom_point(aes(x=Comp3, y=log(FDR), colour = Filter, shape = Referenced)) +
  scale_shape_manual(values=c(8, 20)) +
  theme_classic() +
  geom_hline(aes(yintercept=log(filterRerFdr)), col = "red")

grid.arrange(rerPlotComp1, rerPlotComp2, rerPlotComp3)
```

### PCA on Libs

```{r libs-pca}
# User can choose the number of axis to keep in the PCA
nbPcaAxis <- 3

# Can not perform PCA on qualitative values
filterLibs <- !colnames(libs) %in% c("samplename", "group", "barcode", "index", "total", "spikes")

libsForPca <- libs[filterExtracted, filterLibs]
libsPca <- dudi.pca(libsForPca, center = T, scale = T, scannf = F, nf = nbPcaAxis)
```

```{r libs-pca-summary, dev=c('png', 'svg')}
barplot(libsPca$eig, xlab = "Eigenvalues")
summary(libsPca)
```

```{r libs-pca-lico, dev=c('png', 'svg')}
if (nbClusters != 0) {
  libsPlot <- pcaPlots(libsPca, colorsPcaLi, size, transparency)
} else {
  libsPlot <- pcaPlots(libsPca, size = size, transparency = transparency)
}

# ISSUE: variable name: it is called genes but in the case of libs it is variables
ADEgS(list(libsPlot$genes$ax12, libsPlot$genes$ax32, libsPlot$genes$ax13,
           libsPlot$cells$ax12, libsPlot$cells$ax32, libsPlot$cells$ax13),
  layout = c(2,3))
```

### Diversity among clusters

```{r diversity, dev=c('svg', 'png')}
if (nbClusters != 0) {
  # User should choose which clusters (defined by their color) are to be grouped and compared
  groupeDiversity <- colorsPcaLi %>%
    sapply(function(X) ifelse(X=="grey40", "RightGroup", X)) %>% 
    sapply(function(X) ifelse(X=="darkorange3" || X=="green", "LeftGroup", X)) %>%
    sapply(function(X) ifelse(X=="RightGroup" || X=="LeftGroup", X, "Other")) %>% 
    as.factor()
  
  boxplot(libs$r100l1[filterExtracted]~groupeDiversity)
}
```

### Effect of wells localisation ?

```{r localisation-effect, dev=c('svg','png')}
# User change the plaque specifications
nbWell <- 96
nbColPlaque <- 12


plaqueWellNumber <- matrix(1:nbWell, ncol = nbColPlaque, byrow = T)
coordX <- sapply(1:nbWell, function(X) which(plaqueWellNumber == X, arr.ind = T)[2])
coordY <- sapply(1:nbWell, function(X) which(plaqueWellNumber == X, arr.ind = T)[1])
plaque <- data.frame(wellNumber=c(1:nbWell), x=coordX, y=coordY)

cellTypeNumber <- data.frame(fullName=rownames(genesPca$li), cellName=rownames(genesPca$li)) %>% separate(cellName, c("cellType", "cellNumber"), convert = T) %>% select(-cellType)
cellTypeNumber$fullName %<>% as.character()

plaque <- left_join(plaque, cellTypeNumber, by = c("wellNumber" = "cellNumber"))
plaque %<>% replace_na(list(fullName = "none"))

plaque <- left_join(plaque, data.frame(fullName=names(colorsPcaLi), color=colorsPcaLi), by = "fullName")
plaque$color %<>% as.character()
plaque %<>% replace_na(list(color = "black"))

# Helping visual identification of empty or filtered wells
plaque$pch <- rep(19, nrow(plaque))
plaque$pch[plaque$color=="black"] <- 10

plot(plaque$x, plaque$y, col = plaque$color, ylim = c(min(plaque$y)-0.5, max(plaque$y)+0.5), pch = plaque$pch, cex = 2, axes = F, xlab = "", ylab = "")
text(plaque$x, plaque$y+0.25, labels = plaque$fullName, cex = 0.7, col = plaque$color)
```

Gene counts and TSS discovery
-----------------------------

### Gene count by error code.

```{r gene-count, dev=c('svg', 'png'), fig.height=2.5}
dotsize <- 50
ggplot(libs, aes(x=Group, y=genes)) +
  stat_summary(fun.y=mean, fun.ymin=mean, fun.ymax=mean, geom="crossbar", color="gray") +
  geom_dotplot(aes(fill=Group), binaxis='y', binwidth=1, dotsize=dotsize, stackdir='center') +
  coord_flip() + facet_wrap(~Run)
```

### Gene count per transcript count.

```{r gene-counts-run-plot, dev=c('svg', 'png')}
with(libs, plotNameColor(genes, counts, Run, Well))
```

### Gene counts per C1 run.

```{r gene-counts-run-boxplot, dev=c('svg', 'png')}
#boxplot(data=subset(libs, Error == "0-No Error"), genes ~ Run, ylab="Number of genes detexted (aprox)", main="Comparison of gene detection by run.")
#t.test(data=subset(libs, Error == "0-No Error"), genes ~ Run)
```

Rarefaction (hanabi plot).
--------------------------

FIXME: dark orange !

### Rarefaction

```{r calculate-rarefactions}
# If applicable, restore the previous computation with the command below:
# load("rar.Rda")

# Subsampling sizes: one point for each cell's sample size, plus a few
# points for smaller values, to be able to plot smoothly.

subSampleSizes <- function(X)
  sort(unique(round(c( X
                   , max(X) / 2^(16:0)
                   , (max(X) / 16) * 16:1))))

sampleSizeCounts <- subSampleSizes(colSums(l1))

# Cachig the transpositions, just in case
# it matters for the performance.

t1 <- t(l1)
tg <- t(genes)

# Rarefy at each sampling point.  It takes time !
# In the resluting tables, the columns are a subsampling size, and
# the rows are a cell.

rar1 <- sapply(sampleSizeCounts, function(X) rarefy(t1, X))
rarg <- sapply(sampleSizeCounts, function(X) rarefy(tg, X))

samplesOK <- bed$library %>% factor %>% levels %>% sub('RunA',RunA,.) %>% sub('RunB',RunB,.) # somes samples will be missing if they had no reads.

sampleSizePairs <- subSampleSizes(libs[samplesOK, "properpairs"])

rarefyLong <- function(TABLE, N)
  tapply(TABLE$score, TABLE$library, rarefy, N)

rarU <- sapply(sampleSizePairs, function(N) rarefyLong(bed, N))

# Save the result on the hard drive.  This way, if the commands have been run through knitr,
# the long computations here can be skipped if needed again in an interactive session.
save(rar1, rarg, rarU, sampleSizeCounts, sampleSizePairs,  file="rar.Rda")
```

### TSS discovery

```{r hanabi-TSS, dev=c('svg', 'png')}
hanabiPlot(rar1, sampleSizeCounts, ylab='number of TSS detected', xlab='number of unique molecule counts', main=paste("TSS discovery for", LIBRARY), GROUP=libs$Group)

legend('topleft',legend=levels(libs$Group), col=1:length(levels(libs$Group)), pch=1)
```

### Gene discovery

```{r hanabi-gene, dev=c('svg', 'png')}
hanabiPlot(rarg, sampleSizeCounts, ylab='number of genes detected', xlab='number of unique molecule counts', main=paste("Gene discovery for", LIBRARY), GROUP=libs$Group)

legend('topleft',legend=levels(libs$Group), col=1:length(levels(libs$Group)), pch=1)
```

### Molecule discovery (with raw reads)

````{r hanabi-UMI, dev=c('svg', 'png')}
hanabiPlot(rarU, sampleSizePairs, ylab='number of molecules detected', xlab='number of properly mapped reads', main=paste("Transcript discovery for", LIBRARY), GROUP=libs$Group)

legend('topleft',legend=levels(libs$Group), col=1:length(levels(libs$Group)), pch=1)
```

### Comparison between the runs

(only "No Error" cells)

```{r hanabi-runs, dev=c('svg', 'png')}
#hanabiPlot(rarg[libs$Error=="0-No Error",], sampleSizeCounts, ylab='number of genes detected', xlab='number of unique molecule counts', main=paste("Gene discovery for", LIBRARY), GROUP=libs[libs$Error=="0-No Error", "Run"])

#legend('topleft',legend=levels(libs[libs$Error=="0-No Error", "Run"]), col=1:length(levels(libs[libs$Error=="0-No Error", "Run"])), pch=1)
```


Data load on Zenbu.
===================

FIXME: will do later
FIXME: make idempotent

Accessory functions
-------------------

Uploading to Zenbu collaboration `r ZENBU_COLLAB`.

Ad-hoc wrapper to the shell command `zenbu_upload`

```{r define_zenbuUpload}
zenbuUpload <- function ( ...
                        , URL="http://fantom.gsc.riken.jp/zenbu"
                        , verbose=FALSE
                        , echo=FALSE
                        , stdout=TRUE) {
  zenbu <- 'zenbu_upload'
  url <- c('-url', URL)
  args <- sapply(c(url, ...), shQuote)
  if (verbose == TRUE) print(paste(c(zenbu, args), collapse=' '))
  if (echo    == FALSE) {
      system2(zenbu, args, stdout=stdout)
  } else {
      system2('echo', c(zenbu, args), stdout=stdout)
  }
}
```

To transform a file name in a sample identifier.

```{r define_bedToSample}
bedToSample <- function(BED)
  BED %>%
    sub("RunA", RunA, .) %>%
    sub("RunB", RunB, .) %>%
    sub(".bed", "", .)
```

To produce a string of keywords that will uniquely identify a sample.

TODO: move keywords to the top.

```{r define_sampleDescription}
sampleDescription <- function(BED)
  paste( MOIRAI_FRAGS
       , bedToSample(BED)
       , "what kind of data" # For innstance "CAGEscan_fragments"
       , "a keyword for easy retrieval" # For instance "KnitrUpload"
         # add more keywords as needed
       )
```

To upload CAGEscan fragments, only if they have not yet been uploaded.

```{r define_safeZupload}
safeZupload <- function (BED) {
  # Warning: the "notFound" function ignores queued uploads...
  notFound <- function(BED)
    zenbuUpload( "-list", "-filter", sampleDescription(BED)) %>%
      tail(1) %>%
      grepl ("0 uploads --- 0 featuresources --- 0 experiments --- \\[0 total sources\\]", .)
  if(notFound(BED))
    zenbuUpload( "-file",        paste(MOIRAI_FRAGS, BED, sep="/")
               , "-name",        BED  %>% sub(".bed", "", .)
               , "-assembly",    ASSEMBLY
               , "-desc",        sampleDescription(BED)
               , "-collab_uuid", ZENBU_COLLAB
               , "-singletag_exp"
               , stdout="")
}
```

Functions to add metadata tags.

```{r tag-in-zenbu}
zenbuTag <- function (filter, key, value, mode='add', ...) {
  args <- c('-mdedit', filter, mode, key, value)
  zenbuUpload (args, ...)
}

tagError <- function(BED)
  zenbuTag( sampleDescription(BED)
          , 'cellomics'
          , libs[bedToSample(BED), "Error"] %>% as.character)

tagMeta <- function(BED, TAG)
  zenbuTag( sampleDescription(BED)
          , TAG
          , libs[bedToSample(BED), TAG] %>% as.character)

tagComment <- function (Sample, comment, ...) {
  filter <- paste(MOIRAI_FRAGS, Sample, "C1_CAGEscan_fragments", "KnitrUpload")
  zenbuTag(filter, 'cellomics_comment', comment, ...)
}

#apply(libs[,c('Well', 'Error')],   1, function(X) tagError  (X[1], X[2]))
#apply(libs[,c('Well', 'Comment')], 1, function(X) tagComment(X[1], X[2]))

#keywords need to be stringent: "D12"" matches BED12 !!!
```

### Zenbu uploads and tagging

```{r upload-and-tag}
samplesToUpload <- subset(libs, properpairs > 0, "samplename", drop=T) %>%
  sub(RunA, "RunA", .) %>%
  sub(RunB, "RunB", .) %>%
  sub("$", ".bed", .)

sapply(samplesToUpload, safeZupload)
sapply(samplesToUpload, tagError)
sapply(samplesToUpload, tagMeta, TAG="Group")
sapply(samplesToUpload, tagMeta, TAG="Run")
sapply(samplesToUpload, tagMeta, TAG="row")
sapply(samplesToUpload, tagMeta, TAG="column")
```
