---
title: "<Insert title>"
output: 
  html_document: 
    keep_md: yes
    number_sections: yes
    toc: yes
---

```{r echo=FALSE}
knitr::opts_chunk$set(cache  = TRUE)
knitr::opts_knit$set(verbose = TRUE)
# knitr::opts_chunk$set(dev=c('svg', 'png'))
options(width = 100)
ggplot2::theme_set(ggplot2::theme_bw())
```

```{r library_information_1}
LIBRARY      <- "<Insert library ID>"  # For example, the sequencing run ID.
ZENBU_COLLAB <- "<Insert Zenbu collaboration ID>"
WORKFLOW     <- "<Insert workflow ID>" # For example "OP-WORKFLOW-CAGEscan-short-reads-v2.0"
MOIRAI_STAMP <- "<Insert Moirai run stamp>" # For example "20150803121351"
MOIRAI_PROJ  <- "<Insert Moirai user name>" # For example "Charles"
MOIRAI_USER  <- "<Insert Moirai user name>" # For example "nano-fluidigm"
ASSEMBLY     <- "<Insert name of genome assembly>" # For example "hg38"

# The following is still hardcoded...
# See https://gist.github.com/charles-plessy/9dbc8bc98fb773bf71b6 for hints

GROUP_SHARED <- "/osc-fs_home/scratch/gmtu"
GENE_SYMBOLS <- paste0(GROUP_SHARED, "/annotation/homo_sapiens/gencode-23/gencode.v23.annotation.genes.bed")
ANNOTATION   <- paste0(GROUP_SHARED, "/annotation/homo_sapiens/gencode-23/gencode.v23.annotation.bed")

# In case of a pair of Fludigm C1 CAGE runs, uncomment the lines below and fill
# insert missing information.  Otherwise, delete or leave as it is.
#RunA  <- "<Insert ID of Run A>"
#RunB  <- "<Insert ID of Run B>"
#ctrls <- list( RunA=list(posi="<Insert well ID>", nega="<Insert well ID>")
#             , RunB=list(posi="<Insert well ID>", nega="<Insert well ID>"))
```

<!-- Example title: MiSeq run `r LIBRARY` -->

<Insert title>
==============

 - C1 runs `r RunA` and `r RunB`.
 - Sequenced on MiSeq (`r LIBRARY`)
 
<!-- The lines about cellomics and controls below can be deleted if not using
     C1 CAGE -->

 - Positive control (200 cells): `r ctrls$RunA$posi`, `r ctrls$RunB$posi`.
 - Negative control: `r ctrls$RunA$nega`, `r ctrls$RunB$nega`.
 
Jump directly to the [Analysis](#analysis) section if you are not interested in the details of the data processing.


Data Processing
===============

Link files from MOIRAI.
-----------------------

```{r library_information_2}
BASEDIR      <- "/osc-fs_home/scratch/moirai"
MOIRAI_RUN   <- paste(LIBRARY, WORKFLOW, MOIRAI_STAMP, sep=".")
MOIRAI_BASE  <- paste(BASEDIR, MOIRAI_USER, "project", MOIRAI_PROJ, MOIRAI_RUN, sep="/")
MOIRAI_ID    <- paste(LIBRARY, WORKFLOW, MOIRAI_STAMP, sep=".")
MOIRAI_FRAGS <- paste0(MOIRAI_BASE, "/CAGEscan_fragments")
MOIRAI_BAM   <- paste0(MOIRAI_BASE, "/genome_mapped")
MOIRAI_URL   <- paste0("http://moirai.gsc.riken.jp/", MOIRAI_BASE, "/", MOIRAI_RUN, ".html")

smallCAGEqc::exportInEnv("LIBRARY", "MOIRAI_BASE", "GENE_SYMBOLS", "ANNOTATION")
```

 Intranet link: [`r MOIRAI_RUN`](`r MOIRAI_URL`)

Create a symbolic link from Moirai results to the working directory.

```{r create_link, engine="bash"}
[ -L $LIBRARY.l1.osc.gz ] ||
  ln --symbolic $MOIRAI_BASE/$LIBRARY.l1.osc.gz
```

Annotation and gene symbols.
----------------------------

```{r annotate_bed, engine="bash"}
function osc2bed {
  zcat $1 |
    grep -v \# |
    sed 1d |
    awk '{OFS="\t"}{print $2, $3, $4, "l1", "1000", $5}'
}

function bed2annot {
  bedtools intersect -a $1 -b $ANNOTATION -s -loj |
    awk '{OFS="\t"}{print $1":"$2"-"$3$6,$10}' | 
    bedtools groupby -g 1 -c 2 -o collapse
}

function bed2symbols {
  bedtools intersect -a $1 -b $GENE_SYMBOLS -s -loj |
    awk '{OFS="\t"}{print $1":"$2"-"$3$6,$10}' | 
    bedtools groupby -g 1 -c 2 -o distinct
}

osc2bed $LIBRARY.l1.osc.gz | tee $LIBRARY.l1.bed | bed2annot - > $LIBRARY.l1.annot
bed2symbols $LIBRARY.l1.bed > $LIBRARY.l1.genes
```

Data load and QC in R.
======================

Moirai metadata
---------------

```{r load_R_libraries, message=F}
library(oscR)
library(smallCAGEqc)
stopifnot(
    packageVersion("oscR") >= "0.2.0"
  , packageVersion("smallCAGEqc") > "0.11.2"
)
library(data.table)
library(magrittr)
library(reshape)
library(ggplot2)
library(vegan)
library(plyr)
```

```{r load_libs_from_moirai}
libs <- loadMoiraiStats(processed_data = MOIRAI_BASE)
# Pipe (%>%) to llPostProcess('nano-fluidigm') if appropriate
```

Transcript counts (properly paired)
-----------------------------------

Note that `fread` needs to temporarly save an uncompressed copy of the file
that is being read.  If there is not enough space in the temporary directory
(often `\tmp`),  either uncompress the file first, or point to a larger
directory to contain the temporary files (for example, run `R` with `TMPDIR=.
R`).

Note that _data.tables_ seem to be saved as simple _data frames_ when
closing a session.  After restoring a workspace, one needs to re-convert with the `data.table` command.

```{r load_osc_data}
l1 <- fread.osc(paste0(LIBRARY, ".l1.osc.gz"), dropIdCoords=TRUE)

setnames(l1, colnames(l1) %>% sub('raw.', '', .) %>% sub('.None', '', .) %>% sub('RunA',RunA,.) %>% sub('RunB',RunB,.))

# Be sure that the "libs" and "l1" tables have matching names.
stopifnot(all(colnames(l1) == rownames(libs)))
```

Annotation
----------

```{r annotate_l1}
annot.l1 <- read.table(paste(LIBRARY,'l1','annot',sep='.'), head=F, col.names=c('id', 'feature'), row.names=1)
annot.l1 <- hierarchAnnot(annot.l1)

libs <- cbind(libs, t(rowsum(l1,  annot.l1[,'class']))[rownames(libs),])
```

Gene count
----------

```{r genesymbols_l1}
genesymbols <- read.table(paste(LIBRARY,'l1','genes',sep='.'), col.names=c("cluster","symbol"), stringsAsFactors=FALSE)
rownames(genesymbols) <- genesymbols$cluster

genes <- rowsum(l1, genesymbols$symbol)

libs$genes       <- colSums(genes > 0)
libs$geneSymbols <- countSymbols(genes) # Not the same!
```


Raw reads per molecule (BED12 data)
-----------------------------------

```{r load_fragments}
bed <- loadBED12( libs[libs$counts > 0, "samplename"] %>%
             sub(RunA, "RunA", .) %>%
             sub(RunB, "RunB", .) %>%
             paste0(MOIRAI_FRAGS, "/", ., ".bed"))
```

Cell pictures
-------------

```{r load_fluo_data}
read.fluo <- function(RUN) read.delim( paste0("../imageJ/", RUN, ".txt")
                                     , row.names="cell_id"
                                     , stringsAsFactors = FALSE)
fluo <- rbind(read.fluo(RunA), read.fluo(RunB))
libs <- cbind(libs, subset(fluo,,c("mean_ch2", "bg_mean_ch2", "mean_ch3", "bg_mean_ch3", "Error", "Comment")))
libs$Error <- factor( libs$Error
                    , levels=c("0-No Error", "1-No cell", "2-Debris", "3-OutOfFocus", "4-MultipleCells", "5-Control", "6-Dead"))
```

A hardocoded treshold of 2.5 is used to identify dead cells.  The histogram
below is to check if this value makes sense in this dataset.

```{r dead-cells, dev=c('svg', 'png'), fig.height=2.5}
hist( libs$mean_ch3 - libs$bg_mean_ch3
    , br   = 100
    , main = "Current threshold for identifying dead cells"
    , xlab = "mean_ch3 - bg_mean_ch3" )
deadThresh <- 2.5
abline(v=deadThresh, col="red")

libs[libs$mean_ch3 - libs$bg_mean_ch3 > deadThresh, "Error"] <- "6-Dead"
```

Controls
--------

Some samples with errors were repalced by the positive and negative controls.

```{r flag-controls}
libs[libs$Well %in% ctrls[["RunA"]] & libs$Run == RunA, "Error"]   <- "5-Control"
libs[libs$Well %in% ctrls[["RunB"]] & libs$Run == RunB, "Error"]   <- "5-Control"
libs[libs$Well == ctrls$RunA$posi & libs$Run == RunA, "Comment"] <- "Positive control"
libs[libs$Well == ctrls$RunB$posi & libs$Run == RunB, "Comment"] <- "Positive control"
libs[libs$Well == ctrls$RunA$nega & libs$Run == RunA, "Comment"] <- "Negative control"
libs[libs$Well == ctrls$RunB$nega & libs$Run == RunB, "Comment"] <- "Negative control"
```

cDNA concentration.
-------------------

```{r cDNA_concentration, dev=c('svg', 'png'), fig.height=2.5, message=FALSE, warning=FALSE}
read.pg <- function(RUN)
  paste0("../cDNA_yields/", RUN, ".picogreen.xlsx") %>%
    fldgmPicoGreen("PN 100-6160") %>%
    extract(,"concentration")
libs$Concentration <- c(read.pg(RunA), read.pg(RunB))
fldgmConcentrationPlot(libs)
```

Combined analysis of fluorescence and cDNA concentration.
---------------------------------------------------------

### Array heatmaps.

```{r define_fldgmArrayQCplot}
fldgmArrayQCplot <- function(RUN) fldgmArrayQC(libs[libs$Run==RUN,], RUN)
```

```{r 'runA.arrayQC', dev=c('svg', 'png'), fig.height=2.5}
fldgmArrayQCplot(RunA)
```

```{r 'runB.arrayQC', dev=c('svg', 'png'), fig.height=2.5}
fldgmArrayQCplot(RunB)
```

### Live / dead stain and DNA concentration.

Both runs are plotted together; this explains why there may be two groups, when cDNA
yields differed strongly. 

```{r live-dead, dev=c('svg', 'png')}
with(libs, plotNameColor(Concentration, mean_ch2 - bg_mean_ch2, Error, Well))
with(libs, plotNameColor(Concentration, mean_ch3 - bg_mean_ch3, Error, Well))
```

Richness
--------

```{r calculate-richness}
libs$l1 <- colSums(l1 > 0)
libs$r100l1 <- rarefy(t(l1),100)
```

Analysis
========

```{r richness-concentration, dev=c('svg', 'png')}
with(libs, plotNameColor(Concentration, r100l1, Error, Well))
```

```{r richness-outliers, dev=c('svg', 'png')}
with(subset(libs, Error == "0-No Error"), plotNameColor(Concentration, r100l1, Run, Well))
```

QC barplots
-----------

```{r qc-barplots, dev=c('svg', 'png')}
plotAnnot(libs, 'qc', LIBRARY, libs$Group)
```

Annotation
----------

```{r annotation-barplots, dev=c('svg', 'png')}
plotAnnot(libs, 'counts', LIBRARY, libs$Group)
```

Correlation between runs
------------------------

```{r correlation-heatmap, dev=c('svg', 'png')}
NMF::aheatmap( cor(genes[-1, ])
             , annCol=list(Run=libs$Ru))
```

```{r correlation-heatmap-noerrors, dev=c('svg', 'png')}
# Uncomment if you have a Error column
#NMF::aheatmap( cor(genes[-1, libs$Error == "0-No Error"])
#             , annCol=list(Run=libs[libs$Error == "0-No Error", "Run"]))
```

Gene counts and TSS discovery
-----------------------------

### Gene count by error code.

```{r gene-count, dev=c('svg', 'png'), fig.height=2.5}
dotsize <- 50
ggplot(libs, aes(x=Group, y=genes)) +
  stat_summary(fun.y=mean, fun.ymin=mean, fun.ymax=mean, geom="crossbar", color="gray") +
  geom_dotplot(aes(fill=Group), binaxis='y', binwidth=1, dotsize=dotsize, stackdir='center') +
  coord_flip() + facet_wrap(~Run)
```

### Gene count per transcript count.

```{r gene-counts-run-plot, dev=c('svg', 'png')}
with(libs, plotNameColor(genes, counts, Run, Well))
```

### Gene counts per C1 run.

```{r gene-counts-run-boxplot, dev=c('svg', 'png')}
#boxplot(data=subset(libs, Error == "0-No Error"), genes ~ Run, ylab="Number of genes detexted (aprox)", main="Comparison of gene detection by run.")
#t.test(data=subset(libs, Error == "0-No Error"), genes ~ Run)
```

Rarefaction (hanabi plot).
--------------------------

The purpose of these plots is to ease the comparison of the libraries given that they do not have exactly the same sequencing depth, and to evaluate if extra sequencing depth would be useful.

FIXME: dark orange !

### Rarefaction

```{r calculate-rarefactions}
# If applicable, restore the previous computation with the command below:
# load("rar.Rda")

# Rarefy at with enough sampling points to give a smooth appearance to
# the curves.  It can take time !

rar1 <- hanabi(l1, from = 0)
rarg <- hanabi(genes, from = 0)
rarU <- tapply(bed$score, bed$library, hanabi, from = 0) %>% 
          structure(class = "hanabi") # tapply discards the class !

# Save the result on the hard drive.  This way, if the commands have been run
# through knitr, the long computations here can be skipped if needed again in
# an interactive session.
save(rar1, rarg, rarU, file="rar.Rda")
```

### TSS discovery

```{r hanabi-TSS, dev=c('svg', 'png')}
hanabiPlot( rar1
          , ylab='number of TSS detected'
          , xlab='number of unique molecule counts'
          , main=paste("TSS discovery for", LIBRARY)
          , GROUP=libs$group)
```

### Gene discovery

```{r hanabi-gene, dev=c('svg', 'png')}
hanabiPlot( rarg
          , ylab='number of genes detected'
          , xlab='number of unique molecule counts'
          , main=paste("Gene discovery for", LIBRARY)
          , GROUP=libs$group)
```

### Transcript discovery (with raw reads)

````{r hanabi-UMI, dev=c('svg', 'png')}
hanabiPlot( rarU
          , ylab='number of molecules detected'
          , xlab='number of properly mapped reads'
          , main=paste("Transcript discovery for", LIBRARY)
          , GROUP=libs$Group)
```

### Comparison between the runs

(only "No Error" cells)

```{r hanabi-runs, dev=c('svg', 'png')}
#hanabiPlot(rarg[libs$Error=="0-No Error",], sampleSizeCounts, ylab='number of genes detected', xlab='number of unique molecule counts', main=paste("Gene discovery for", LIBRARY), GROUP=libs[libs$Error=="0-No Error", "Run"])

#legend('topleft',legend=levels(libs[libs$Error=="0-No Error", "Run"]), col=1:length(levels(libs[libs$Error=="0-No Error", "Run"])), pch=1)
```

Note that the horizontal axis is not the same in every plot:

- For _TSS_ or _Gene discovery_, the horizontal axis is unique molecule counts, that is, transcripts. The purpose of these plots is to show how easy it would be to detect more TSS or genes if we would sequence more transcripts.

- For _Transcript discovery_, the horizontal axis is number of raw reads. The purpose of this plot is to show how possible it is to sequence more transcripts by increasing the number of raw reads.


Data load on Zenbu.
===================

FIXME: will do later
FIXME: make idempotent

Accessory functions
-------------------

Uploading to Zenbu collaboration `r ZENBU_COLLAB`.

Ad-hoc wrapper to the shell command `zenbu_upload`

```{r define_zenbuUpload}
zenbuUpload <- function ( ...
                        , URL="http://fantom.gsc.riken.jp/zenbu"
                        , verbose=FALSE
                        , echo=FALSE
                        , stdout=TRUE) {
  zenbu <- 'zenbu_upload'
  url <- c('-url', URL)
  args <- sapply(c(url, ...), shQuote)
  if (verbose == TRUE) print(paste(c(zenbu, args), collapse=' '))
  if (echo    == FALSE) {
      system2(zenbu, args, stdout=stdout)
  } else {
      system2('echo', c(zenbu, args), stdout=stdout)
  }
}
```

Helper functions

```{r define_helperFunctions}
bedToSample <- function(BED)
  BED %>%
    sub("RunA", RunA, .) %>%
    sub("RunB", RunB, .) %>%
    sub(".bed", "", .)

# To produce a string of keywords that will uniquely identify a sample.
# TODO: move keywords to the top.

sampleDescription <- function(BED)
  paste( MOIRAI_FRAGS
       , bedToSample(BED)
       , "what kind of data" # For innstance "CAGEscan_fragments"
       , "a keyword for easy retrieval" # For instance "KnitrUpload"
         # add more keywords as needed
       )

CSfragmentLocation <- function(BED)
  paste0(MOIRAI_FRAGS, "/", BED, ".bed")

CSuploadCommand <- function(BED)
  paste( CSfragmentLocation(BED)
       , bedToSample(BED)
       , sampleDescription(BED)
       , ""
       , sep = "\t")
```

```{r createUploadFile}
levels(bed$library) %>%
  CSuploadCommand %>%
  writeLines(paste0(LIBRARY, ".zUpload.tsv"))
```

To upload CAGEscan fragments, only if they have not yet been uploaded.

```{r zenbuFileListUpload}
zenbuUpload( "-assembly", ASSEMBLY
           , "-filelist", paste0(LIBRARY, ".zUpload.tsv")
           , "-collab_uuid", ZENBU_COLLAB
           , "-singletag_exp")
```

Functions to add metadata tags.

```{r tag-in-zenbu}
zenbuTag <- function (filter, key, value, mode='add', ...) {
  args <- c('-mdedit', filter, mode, key, value)
  zenbuUpload (args, ...)
}

tagError <- function(BED)
  zenbuTag( sampleDescription(BED)
          , 'cellomics'
          , libs[bedToSample(BED), "Error"] %>% as.character)

tagMeta <- function(BED, TAG)
  zenbuTag( sampleDescription(BED)
          , TAG
          , libs[bedToSample(BED), TAG] %>% as.character)

tagComment <- function (Sample, comment, ...) {
  filter <- paste(MOIRAI_FRAGS, Sample, "C1_CAGEscan_fragments", "KnitrUpload")
  zenbuTag(filter, 'cellomics_comment', comment, ...)
}

#apply(libs[,c('Well', 'Error')],   1, function(X) tagError  (X[1], X[2]))
#apply(libs[,c('Well', 'Comment')], 1, function(X) tagComment(X[1], X[2]))

#keywords need to be stringent: "D12"" matches BED12 !!!
```

### Zenbu uploads and tagging

```{r upload-and-tag}
samplesToUpload <- subset(libs, properpairs > 0, "samplename", drop=T) %>%
  sub(RunA, "RunA", .) %>%
  sub(RunB, "RunB", .) %>%
  sub("$", ".bed", .)

sapply(samplesToUpload, tagError)
sapply(samplesToUpload, tagMeta, TAG="Group")
sapply(samplesToUpload, tagMeta, TAG="Run")
sapply(samplesToUpload, tagMeta, TAG="row")
sapply(samplesToUpload, tagMeta, TAG="column")
```
